### PR TITLE
user: improve register and forgot password

### PR DESCRIPTION
**Pull request by Garfield-fr**
* Adds beautifulsoup4 package for test.
* Disables account rest api endpoint.
* Adds validations on the change password api.
* Updates password on the interface test information.
* Adds the possibility to put a message in the cache
  and display it on the html page.
* Changes of the API response on invalid login.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -410,6 +410,10 @@ vendors = "rero_ils.modules.vendors.mappings"
 operation_logs = "rero_ils.modules.operation_logs.es_templates:list_es_templates"
 rero_ils = "rero_ils.es_templates:list_es_templates"
 
+
+[tool.poetry.group.dev.dependencies]
+beautifulsoup4 = "^4.11.1"
```
Comment (zannkukai) [inline pyproject.toml:415]: Youhou è  ! a new dependencies 🥳

File: rero_ils/accounts_views.py [hunk 2]
```diff
@@ -46,6 +51,35 @@ def user_exists(email):
...
+                         ' case letter, one lower case letter'
+                         ' and one number.'),
+            'invalid_special_char': _('The password must contain at least'
+                                      ' one upper case letter, one lower'
+                                      ' case letter, one number and one'
+                                      ' special character.')
+        }
+        message = messages['invalid_special_char'] if special_char \
+            else messages['invalid']
+        raise ValidationError(message)
```
Comment (zannkukai) [inline rero_ils/accounts_views.py:74]: Didn't you already create a validation password method ? Why don't use it .... I think I miss something :s

File: rero_ils/accounts_views.py [hunk 3]
```diff
@@ -71,23 +104,51 @@ def post(self, **kwargs):
         self.login_user(user)
         return self.success_response(user)
 
+    def verify_login(self, user, password=None, **kwargs):
+        """Verify the login via password."""
+        if not user.password or not verify_and_update_password(password, user):
+            _abort(_('INVALID_USER_OR_PASSWORD'))
```
Comment (zannkukai) [inline rero_ils/accounts_views.py:110]: You mark the string `INVALID_USER_OR_PASSWORD` for translation ? Why use a snake case string ?
maybe it should be `get_message('INVALID_USER_OR_PASSWORD')[0]` ?

File: rero_ils/accounts_views.py [hunk 4]
```diff
@@ -139,3 +199,15 @@ def post(self, args):
...
+class DisabledAuthApiView(MethodView):
+    """Disabled Account rest auth api."""
+
+    def post(self, **kwargs):
+        """Post Data."""
+        abort(404)
+
+    def get(self, **kwargs):
+        """Get Data."""
+        abort(404)
```
Comment (zannkukai) [inline rero_ils/accounts_views.py:213]: Suggestion : You could write 
```python
    def post(self, **kwargs):
        """Post Data."""
        abort(404)

    get = post
```
in this case, the GET and POST method are always the same.

File: rero_ils/modules/patrons/api.py [hunk 5]
```diff
@@ -236,7 +236,8 @@ def load(cls, data):
     def removeUserData(cls, data):
         """Remove the user data."""
         data = deepcopy(data)
-        profile_fields = User.profile_fields + ['username', 'email']
+        profile_fields = User.profile_fields + \
+            ['username', 'email', 'password']
```
Comment (zannkukai) [inline rero_ils/modules/patrons/api.py:240]: Oh yes. A future rebase conflict 🤪

File: rero_ils/modules/users/validators.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+"""Password validator."""
+
+from wtforms import ValidationError
+
+from rero_ils.modules.utils import PasswordValidatorException, \
+    password_validator
+
+
+class PasswordValidator(object):
```
Comment (zannkukai) [inline rero_ils/modules/users/validators.py:26]: With Python3, it's recommended to not inherit from `object` --> just write `class PasswordValidator:`

File: rero_ils/modules/users/validators.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    """Password Validator."""
+
+    def __init__(self, length=8, special_char=False):
+        """Password validator class initializer."""
+        self.length = length
+        self.special_char = special_char
+
+    def __call__(self, form, field):
```
Comment (zannkukai) [inline rero_ils/modules/users/validators.py:34]: missing docstring about params and possible `raise` statements

File: rero_ils/modules/users/validators.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        """Call."""
+        field_length = field.data and len(field.data) or 0
+        messages = {
+            'length': field.ngettext(
+                'Field must be at least %(min)d character long.',
+                'Field must be at least %(min)d characters long.',
+                self.length)
+        }
```
Comment (zannkukai) [inline rero_ils/modules/users/validators.py:42]: Seems a little bit overkill to declare a dict with only 1 key. You could simply use `field.ngettext` into the belowed `if` statement --> 
```python
raise ValidationError(field.ngettext(
      'Field must be at least %(min)d character long.',
      'Field must be at least %(min)d characters long.',
      self.length
))
```
Additionnaly, it should be translated. Isn't it ?

File: rero_ils/theme/templates/rero_ils/_info_test.html [hunk 9]
```diff
@@ -20,31 +20,31 @@
   <h3>Test RERO ILS</h3>
   <p>As an anonymous user</p>
   <p>As a patron, login with:</br>
-  reroilstest+giulia@gmail.com / 123456</p>
+  reroilstest+giulia@gmail.com / Pw123456</p>
```
Comment (zannkukai) [inline rero_ils/theme/templates/rero_ils/_info_test.html:23]: I love security 🤣

File: rero_ils/accounts_views.py [hunk 10]
```diff
@@ -46,6 +53,25 @@ def user_exists(email):
         raise ValidationError(_('INVALID_USER_OR_PASSWORD'))
 
 
+def validate_password(password):
+    """Validate the password."""
+    length = current_app.config.get('RERO_ILS_PASSWORD_MIN_LENGTH', 8)
+    special_char = current_app.config.get('RERO_ILS_PASSWORD_SPECIAL_CHAR')
+    if len(password) < length:
+        raise ValidationError(
+            _('Field must be at least %(min)d characters long.', min=length))
```
Comment (zannkukai) [inline rero_ils/accounts_views.py:62]: As password length check is already include into `password_validator`, why this test here ?

File: rero_ils/modules/users/validators.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        self.special_char = special_char
+
+    def __call__(self, form, field):
+        """Call.
+
+        :param form: the current form.
+        :param field: the password field.
+        :raise ValidationError: If the password isn't valid.
+        """
+        field_length = field.data and len(field.data) or 0
```
Comment (zannkukai) [inline rero_ils/modules/users/validators.py:41]: Seems strange... `field.data` should return boolean. So your test `field.data and len(field.data)` will compare a boolean and an integer ; this integer will be casted to boolean. So `field_length` should return `True|False` (impossible to return 0).
Next line, you compare `field_length < self.length`, so the boolean value is now casted to int (0|1).

A solution should be to use the ugly python ternary operator : `field_length = len(field.data) if field.data else 0`

File: rero_ils/modules/users/validators.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        if field_length < self.length:
+            raise ValidationError(field.ngettext(
+                'Field must be at least %(min)d character long.',
+                'Field must be at least %(min)d characters long.',
+                self.length
+            ) % {"min": self.length})
```
Comment (zannkukai) [inline rero_ils/modules/users/validators.py:47]: same remark. The length validation is already included into `password_validator`

**Review by rerowep — APPROVED** [review]

**Review by zannkukai — APPROVED** [review]

**Pull request merged**